File: language-support/js/daml-ledger/src/model/exercise_command.ts
// Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

import * as ledger from '..';

/**
 * @interface ExerciseCommand
 * @memberof ledger
 */
export interface ExerciseCommand {
    /**
     * @member {ledger.Identifier} templateId
     * @memberof ledger.ExerciseCommand
     * @instance
     */
    templateId: ledger.Identifier
    /**
     * @member {string} choice
     * @memberof ledger.ExerciseCommand
     * @instance
     */
    choice: string
    /**
     * @member {string} contractId
     * @memberof ledger.ExerciseCommand
     * @instance
     */
    contractId: string
    /**
     * @member {ledger.Value} argument
     * @memberof ledger.ExerciseCommand
     * @instance
     */
    argument: ledger.Value
}

File: canton/community/ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/scenario/ExerciseCommand.scala
// Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
package com.daml.http.perf.scenario

import io.gatling.core.Predef._
import io.gatling.http.Predef._

@SuppressWarnings(Array("org.wartremover.warts.NonUnitStatements"))
class ExerciseCommand extends Simulation with SimulationConfig {

  private val createCommand = s"""{
  "templateId": "Iou:Iou",
  "payload": {
    "issuer": "$aliceParty",
    "owner": "$aliceParty",
    "currency": "USD",
    "amount": "9.99",
    "observers": []
  }
}"""

  private val exerciseCommand = s"""{
    "templateId": "Iou:Iou",
    "contractId": "$${contractId}",
    "choice": "Iou_Transfer",
    "argument": {
        "newOwner": "$aliceParty"
    }
  }"""

  private val numberOfRuns = 2000
  private val createRequest = http("CreateCommand")
    .post("/v1/create")
    .body(StringBody(createCommand))

  private val exerciseRequest =
    http("ExerciseCommand")
      .post("/v1/exercise")
      .body(StringBody(exerciseCommand))

  private val scn = scenario("ExerciseCommandScenario-Create")
    .repeat(numberOfRuns / defaultNumUsers)(exec(createRequest.silent)) // populate the ACS

  private val queryScn = scenario("ExerciseCommandScenario-QueryAndExercise")
    .repeat(1)(
      exec(
        // retrieve all contractIds
        http("GetACS")
          .get("/v1/query")
          .check(status.is(200), jsonPath("$.result[*].contractId").findAll.saveAs("contractIds"))
          .silent
      )
    )
    .foreach("${contractIds}", "contractId")(exec(exerciseRequest))

  setUp(
    scn
      .inject(atOnceUsers(defaultNumUsers))
      .andThen(queryScn.inject(atOnceUsers(1)))
  ).protocols(httpProtocol)
}
